module_analysis/models/ir_module_module.py

```diff
@@ -0,0 +1,180 @@
+            },
+            '.xml': {
+                'code': 'xml_code_qty',
+            },
+            '.js': {
+                'code': 'js_code_qty',
+            },
+            '.css': {
+                'code': 'css_code_qty',
+            },
+        }
+
+    @api.model
+    def _get_clean_analyse_values(self):
+        """List of fields to unset when a module is uninstalled"""
+        return {
+            'author_ids': [6, 0, []],
+            'module_type_id': False,
+            'python_code_qty': False,
+            'xml_code_qty': 0,
+            'js_code_qty': 0,
+            'css_code_qty': 0,
+        }
+
+    # Overload Section
+    @api.model
+    def update_list(self):
+        res = super().update_list()
+        if self.env.context.get('analyse_installed_modules', False):
+            self.search([('state', '=', 'installed')]).button_analyse_code()
+        return res
+
+    @api.multi
+    def write(self, vals):
+        res = super().write(vals)
+        if vals.get('state', False) == 'installed':
+            self.button_analyse_code()
+        elif vals.get('state', False) == 'uninstalled'\
+                and 'module_analysis' not in [x.name for x in self]:
+            self.write(self._get_clean_analyse_values())
+        return res
+
+    # Public Section
+    @api.multi
+    def button_analyse_code(self):
+        IrModuleAuthor = self.env['ir.module.author']
+        IrModuleTypeRule = self.env['ir.module.type.rule']
+        rules = IrModuleTypeRule.search([])
+
+        cfg = self.env['ir.config_parameter']
+        val = cfg.get_param('module_analysis.exclude_directories', '')
+        exclude_directories = [x.strip() for x in val.split(',') if x.strip()]
+        val = cfg.get_param('module_analysis.exclude_files', '')
+        exclude_files = [x.strip() for x in val.split(',') if x.strip()]
+
+        for module in self:
+            _logger.info("Analysing Code for module %s ..." % (module.name))
+
+            # Update Authors, based on manifest key
+            authors = []
+            if module.author and module.author[0] == '[':
+                author_txt_list = safe_eval(module.author)
+            else:
+                author_txt_list = module.author.split(',')
+
+            author_txt_list = [x.strip() for x in author_txt_list]
+            author_txt_list = [x for x in author_txt_list if x]
+            for author_txt in author_txt_list:
+                authors.append(IrModuleAuthor._get_or_create(author_txt))
+
+            author_ids = [x.id for x in authors]
+            module.author_ids = author_ids
+
+            # Update Module Type, based on rules
+            module_type_id = rules._get_module_type_id_from_module(module)
+            module.module_type_id = module_type_id
+
+            # Get Path of module folder and parse the code
+            module_path = get_module_path(module.name)
+
+            # Get Files
+            analysed_datas = self._get_analyse_data_dict()
```

Comment:
it's wrong to write "datas" with s; it looks like Odoo SA code. So like many latin words, data is already the plural form. Would you want to use the singular singular, you would write datum. There are also 4 other occurences of "datas" in the subsequent 15 lines. https://www.learnenglish.de/mistakes/data.html